runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,544 @@
+    // the result and timing in some way to higher-layer subsystems;
+    TerminatedToFreeze,
+    DroppedFromBankForks,
+    // scheduler will be restarted without being returned to pool in order to reuse it immediately.
+    ReinitializedForRecentBlockhash,
+}
+
+pub type InstalledSchedulerBox = Box<dyn InstalledScheduler<DefaultScheduleExecutionArg>>;
+
+#[derive(Clone, Debug)]
+pub struct SchedulingContext {
+    mode: SchedulingMode,
+    // Intentionally not using Weak<Bank> for performance reasons
+    bank: Arc<Bank>,
+}
+
+impl WithSchedulingMode for SchedulingContext {
+    fn mode(&self) -> SchedulingMode {
+        self.mode
+    }
+}
+
+impl SchedulingContext {
+    pub fn new(mode: SchedulingMode, bank: Arc<Bank>) -> Self {
+        Self { mode, bank }
+    }
+
+    pub fn bank(&self) -> &Arc<Bank> {
+        &self.bank
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.bank().slot()
+    }
+
+    pub fn log_prefix(context: Option<&Self>, scheduler_id: SchedulerId) -> String {
+        const BITS_PER_HEX_DIGIT: usize = 4;
+
+        format!(
+            "sch_{:0width$x}{}: ",
+            scheduler_id,
+            context
+                .map(|c| format!("(slot:{}/mode:{:?})", c.slot(), c.mode()))
+                .unwrap_or_else(|| "(?)".into()),
+            width = SchedulerId::BITS as usize / BITS_PER_HEX_DIGIT,
+        )
+    }
+}
+
+// tiny wrapper to ensure to call wait_for_termination() via ::drop() inside
+// BankForks::set_root()'s pruning.
+#[derive(Clone)]
+pub struct BankWithScheduler {
+    inner: Arc<BankWithSchedulerInner>,
+}
+
+#[derive(Debug)]
+pub struct BankWithSchedulerInner {
+    bank: Arc<Bank>,
+    scheduler: InstalledSchedulerRwLock,
+}
+pub type InstalledSchedulerRwLock = RwLock<Option<InstalledSchedulerBox>>;
+
+#[allow(clippy::declare_interior_mutable_const)]
+pub const NO_INSTALLED_SCHEDULER_RW_LOCK: InstalledSchedulerRwLock = RwLock::new(None);
+
+impl BankWithScheduler {
+    pub(crate) fn new(bank: Arc<Bank>, scheduler: Option<InstalledSchedulerBox>) -> Self {
+        Self {
+            inner: Arc::new(BankWithSchedulerInner {
+                bank,
+                scheduler: RwLock::new(scheduler),
+            }),
+        }
+    }
+
+    #[cfg(any(test, feature = "test-in-workspace"))]
+    pub fn new_for_test(bank: Arc<Bank>, scheduler: Option<InstalledSchedulerBox>) -> Self {
+        Self::new(bank, scheduler)
+    }
+
+    pub fn new_without_scheduler(bank: Arc<Bank>) -> Self {
+        Self::new(bank, None)
+    }
+
+    pub fn bank_cloned(&self) -> Arc<Bank> {
+        self.bank().clone()
+    }
+
+    pub fn bank(&self) -> &Arc<Bank> {
+        &self.inner.bank
+    }
+
+    // don't indefintely lend out scheduler refs to avoid accidental mix because scheduler should
+    // strictly tied to bank
+    pub fn with_scheduler_lock(&self, callback: impl FnOnce(&InstalledSchedulerRwLock)) {
+        callback(&self.inner.scheduler)
+    }
```

Comment:
> I was under the impression, that this kind of wrapping is only necessary when you need to perform some kind of adjustment after the callback.

yeah, I'm aware of these common use cases as well. additionally, i think this closure pattern is used like this to constrain references in rust: https://doc.rust-lang.org/stable/std/thread/struct.LocalKey.html#method.with

> In particular, you can not return anything from the callback here. 

yeah, i just omitted that nicety due to laziness and no need right now.

Reply:
> What does this give you over
> 
> ```rust
> pub fn scheduler(&self) -> &InstalledSchedulerRwLock {
>     &self.inner.scheduler
> }
> ```

with the closure api, I _originally_ thought the following mistake can be made harder to occur:


```rust
let scheduler = bank1.scheduler();
/// 100 lines later
bank2.register_tick(hash, scheduler); // bad!

/// compared to this

bank2.with_scheduler(|scheduler| bank1.register_tick(hash, scheduler) // bad!
```

that's because there's no `scheduler` to bind to a temporary variable with `with_scheduler`.

as i said _originally_, I have found better solution: 3d7c60fa5ab9f95cd454a175ef47068beb9ca2b7

Reply:
> > What does this give you over
> > ```rust
> > pub fn scheduler(&self) -> &InstalledSchedulerRwLock {
> >     &self.inner.scheduler
> > }
> > ```
> 
> with the closure api, I _originally_ thought the following mistake can be made harder to occur:
> 
> ```rust
> let scheduler = bank1.scheduler();
> /// 100 lines later
> bank2.register_tick(hash, scheduler); // bad!
> 
> /// compared to this
> 
> bank2.with_scheduler(|scheduler| bank1.register_tick(hash, scheduler) // bad!
> ```
> 
> that's because there's no `scheduler` to bind to a temporary variable with `with_scheduler`.

One can still do this:

```rust
bank1.with_scheduler(|scheduler1| {
  bank2.with_scheduler(|scheduler2| {
    /// 100 lines later
    scheduler2.register_tick(hash);
    /// 100 lines later
    scheduler1.register_tick(hash);
  }
}
```

You did say "harder", so, yes it is headers :)
But it similarly makes it harder to write code that needs to operate on several values.
I am a bit skeptical that it is a good API design instrument for a case like this.
The API user can still do whatever they want, so the safety is not really there.
I doubt that the fact that something is inconvenient would make the user think, "Am I doing it wrong?  Maybe this is unsafe?"

> as i said _originally_, I have found better solution: [3d7c60f](https://github.com/solana-labs/solana/commit/3d7c60fa5ab9f95cd454a175ef47068beb9ca2b7)

This direct solution is 100% better.  It is simpler and to the point :)